### DISCUSSION TITLE
OpenSSL vs LibreSSL

### DISCUSSION THREAD

**Discussion by agucova** [Homebrew/discussions#3258] [Tap maintenance and Homebrew development]
Hi, I've been reading on the decisions by the Homebrew community to depend most formulae on OpenSSL over LibreSSL, and explicitly disallowing LibreSSL as an alternative crypto backend on core formulae (see [#18779](https://github.com/Homebrew/homebrew-core/pull/18779)  and  [#26835](https://github.com/Homebrew/homebrew-core/pull/26835)).

As far as I could gather, Apple has mostly made the change to LibreSSL. While I'm aware that this is a contentious issue for many reasons, I wanted to ask what were the reasons that went behind the initial decision, given that I couldn't find any public records of the discussion.

PD: I'm not sure if this is the right category for this.

**Comment by SMillerDev** [comment]
@Homebrew/maintainers anyone who remembers this?

I'd guess that with the removal of options the analytics showed that people preferred openssl.

**Comment by Bo98** [comment]
This was long before my time, but I can answer based on a theoretical question of switching SSL libraries tomorrow:

* Switching SSL libraries is an enormous task. Everything in the dependency graph should ideally use the same SSL library, and some of them are massive, given how widespread SSL library usage is and how they can be dependencies of other key dependencies. This is part of the reason we haven't migrated to OpenSSL 3 yet (alongside build issues on older macOS versions), but we will eventually.
  - OpenSSL coming first will likely have greatly helped it become the default - Homebrew is older than LibreSSL!
* OpenSSL makes ABI breaking updates much much less often. When an ABI breaking update is made, all dependents must be rebuilt at the same time as the update being shipped, or both versions continue to be offered for a migration period (but we'd still ideally migrate entity dependency graphs at a time to prevent mixed version usage). The number of OpenSSL dependents is so enormous that having to do this task regularly would be a nightmare. Every few years on a major OpenSSL update (like 1.1 -> 3) is enough.
  - This point is less of an issue for Apple because unlike the old OpenSSL 0.9.8 they shipped, Apple don't allow anyone to link against their LibreSSL (and even then, they still sometimes ship several ABI versions of LibreSSL with macOS for even internal compatibility).
